spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,203 @@
+the packet.
+
+Given this, the flow would be:
+
+1. User/module submits a send packet on the `source` chain, along with some tokens and fee information on how to distribute them
+2. Relayer submits `OnReceivePacket` on the `destination` chain. Along with this message, the *forward relayer* will submit a `payToOnSource` address where payment should be sent.
+3. Destination application includes this `payToOnSource` in the acknowledgement (there are multiple approaches to discuss below)
+4. Relayer submits `OnAcknowledgement` which provides the *return relayer* address on the source chain, along with the `payToOnSource` address
+5. Source application can distribute the tokens escrowed in (1) to both the *forward* and the *return* relayers.
+
+Alternate flow:
+
+1. User/module submits a send packet on the `source` chain, along with some tokens and fee information on how to distribute them
+2. Relayer submits `OnTimeout` which provides their address on the source chain
+3. Source application can distribute the tokens escrowed in (1) to this relayer, and potentially return extra tokens to the original packet sender.
+
+### Fee details
+
+For an example implementation in the Cosmos SDK, we consider 3 potential fee payments, which may be defined. Each one may be
+paid out in a different token. Imagine a connection between IrisNet and the Cosmos Hub. They may define:
+
+- ReceiveFee: 0.003 channel-7/ATOM vouchers (ATOMs already on IrisNet via ICS20)
+- AckFee: 0.001 IRIS
+- TimeoutFee: 0.002 IRIS
+
+Ideally the fees can easily be redeemed in native tokens on both sides, but relayers may select others. In this example, the relayer collects a fair bit of IRIS, covering it's costs there and more. It also collects channel-7/ATOM vouchers from many packets. After relaying a few thousand packets, the account on the Cosmos Hub is running low, so the relayer will send those channel-7/ATOM vouchers back over channel-7 to it's account on the Hub to replenish the supply there. 
+
+The sender chain will escrow 0.003 channel-7/ATOM and 0.002 IRIS. In the case that a forward relayer submits the `MsgRecvPacket` and a reverse relayer submits the `MsgAckPacket`, the forward relayer is reqarded 0.003 channel-7/ATOM and the reverse relayer is rewarded 0.001 IRIS. In the case where the packet times out, the timeout relayer receives 0.002 IRIS and 0.003 channel-7/ATOM is refunded to the original fee payer.
+
+The logic involved in collecting fees from users and then paying it out to the relevant relayers is encapsulated by a separate fee module and may vary between implementations. However, all fee modules must implement a uniform interface such that the ICS-4 handlers can correctly pay out fees to the right relayers, and so that relayers themselves can easily determine the fees they can expect for relaying a packet.
+
+### Fee Module Contract
+
+```typescript
+function PayFee(packet: Packet, forward_relayer: string, reverse_relayer: string) {
```

Comment:
I don't understand if we want to enforce token fee escrowing as part of packet sending. Do we need an additional function that can be called as part of sending packet to pass tokens to fee module, or that logic is expected to be implemented as part of IBC application?

Reply:
I think there are two design decisions.

1. How the fee module works?
2. Where we call in the fee module? (From ICS-4 directly? From the apps (which can be implemented as an IbcHandler decorator, or just a function call)

I am more interested in spec-ing (1) first and when that is clear, decide where it fits (2) - which requires much more investigation at the code level. I think we can define all functionality of (1) without even looking at Go. And then it limits the design space we discuss later

Reply:
Thinking more (and looking at pseudocode below), it seems we *should* specify the *interface* for storing the fee for a packet. Whether this is called from ICS-4 in send-packet or in the application can be discussed separately. But the interface on the fee handler can be specified.

`function SetPacketFee(packet: Packet, receiveFee: Fee, ackFee: Fee, timeoutFee: Fee)`

And then define it may be called once, atomically with committing the packet. If it is not called (or called with 0) then the packet is un-incentivized, which is a valid state.

Reply:
I agree we should standardize `SetPacketFee` interface.

> Whether this is called from ICS-4 in send-packet or in the application can be discussed separately. But the interface on the fee handler can be specified.

I'm a little resistant to doing this in ICS-4 in send-packet. Ideally, I'd like this spec to be a natural fit for most custom fee modules.

The usecase you've mentioned above works really well for the default case where sender pays for the packet flow.

However, consider the case of the altruistic module funds packets on a particular channel.
Here, the module does not escrow some additional funds on each packet send. Instead, the funds have been escrowed in advance for the purpose of paying for future packets. Having a `SetPacketFee` being called in ICS-4 handler here would be unnatural. Sure, we could just set 0 fee in ICS-4 and then have the fees come from the previously created escrow account, but this would of course require the ability to set more than once. In this case, there could be no explicit `SetPacketFee` at all, as the fund could just release some constant amount from its escrow on `PayFee` and `PayTimeoutFee` callbacks.

> And then define it may be called once, atomically with committing the packet. If it is not called (or called with 0) then the packet is un-incentivized, which is a valid state.

For reasons I've stated above, I support the ability to call `SetPacketFee` multiple times with each subsequent call adding onto the previous escrow.

I think `SetPacketFee` should be an open function that can be called into by IBC applications, third-party users, DAOs, governance, whatever. And they should all be able to parallelly incentivize a particular packet without interfering with each other. The relayer will reap collective rewards from all incentivizers. I don't believe this is hard to do because of the way we've split responsibilities in this spec. It's a natural usecase of the functions in this spec.

Reply:
Nice argument.

If we allow `SetPacketFee` to be called multiple times, we need a new name.

`IncrementPacketFee`? `EscrowPacketFee`? 

Something that makes clear it is additive.